### PR TITLE
[Enhancement] Add ONNX passes support

### PR DESCRIPTION
**Pull request by grimoire**
## Motivation

This PR adds custom onnx passes support. And two template passes has been added.

## Modification

`_jit_pass_merge_shape_concate`: merge shape concate

origin:
Screenshot from 2022-04-22 16-12-04
new:
Screenshot from 2022-04-22 16-12-20

`_jit_pass_onnx_peephole`: 
- reshape->reshape => reshape
- cast(dtype n)->cast(dtype n) => cast(dtype n)

origin:
Screenshot from 2022-04-22 16-10-50
new:
Screenshot from 2022-04-22 16-10-58

`_jit_pass_flatten_cls_head`

origin:
Screenshot from 2022-05-11 16-42-49
new:
Screenshot from 2022-05-11 16-41-57

## Use cases (Optional)

```python

    def custom_passes(graph, params_dict, torch_out):
        ts_optimizer._jit_pass_merge_shape_concate(graph)
        return graph, params_dict, torch_out

    with RewriterContext({},
                         backend='onnxruntime',
                         onnx_custom_passes=custom_passes):
        torch.onnx.export(
            model,
            inputs,
            new_onnx_path,
            input_names=input_names,
            output_names=output_names)
```

### CODE REVIEW

**Comment by RunningLeon** [comment]
Could we add unit test for each optimize pass

File: tests/test_apis/test_onnx_passes.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (RunningLeon) [inline tests/test_apis/test_onnx_passes.py:1]: binary file `mmdeploy/backend/ncnn/onnx2ncnn` should be removed from git commit.

File: tests/test_apis/test_onnx_passes.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+import tempfile
+from typing import Any, List, Tuple
```
Comment (RunningLeon) [inline tests/test_apis/test_onnx_passes.py:3]: It's weird that all three unit tests failed on my machine:
```
=================================================== FAILURES ===================================================
___________________________________________ test_merge_shape_concate ___________________________________________

    def test_merge_shape_concate():
        pytest.importorskip('mmdeploy.backend.torchscript.ts_optimizer.onnx')
    
        try:
            from mmdeploy.backend.torchscript import ts_optimizer
            opt_pass = ts_optimizer.onnx._jit_pass_merge_shape_concate
        except ImportError:
            pytest.skip('pass not found.')
    
        def _optimize_onnx(graph, params_dict, torch_out):
            opt_pass(graph)
            return graph, params_dict, torch_out
    
        class TestModel(nn.Module):
    
            def __init__(self):
                super().__init__()
    
            def forward(self, x):
                return x.new_zeros(x.shape[-2:])
    
        model = TestModel()
        x = torch.rand(1, 3, 4, 8)
    
        with RewriterContext({}, onnx_custom_passes=_optimize_onnx):
            torch.onnx.export(
                model,
                x,
                onnx_file,
                input_names=['input'],
                output_names=['output'],
                dynamic_axes=dict(input=[2, 3]))
    
        onnx_model = onnx.load(onnx_file)
        graph = onnx_model.graph
        nodes = graph.node
        shape_idx = 0
        for n in nodes:
            if n.op_type != 'Shape':
                shape_idx += 1
            else:
                break
    
        assert shape_idx < len(nodes)
>       assert nodes[shape_idx + 1].op_type == 'Gather'
E       AssertionError: assert 'Constant' == 'Gather'
E         - Gather
E         + Constant

tests/test_apis/test_onnx_passes.py:68: AssertionError
________________________________________________ test_peephole _________________________________________________

    def test_peephole():
        pytest.importorskip('mmdeploy.backend.torchscript.ts_optimizer.onnx')
    
        try:
            from mmdeploy.backend.torchscript import ts_optimizer
            opt_pass = ts_optimizer.onnx._jit_pass_onnx_peephole
        except ImportError:
            pytest.skip('pass not found.')
    
        def _optimize_onnx(graph, params_dict, torch_out):
            opt_pass(graph)
            return graph, params_dict, torch_out
    
        class TestModel(torch.nn.Module):
    
            def __init__(self):
                super().__init__()
    
            def forward(self, x):
    
                x = x.int()
                x = x.int()
                x = x.float()
    
                x = x.view(10, -1)
                y = x.view(2, -1)
                z = x.view(3, -1)
    
                return y, z
    
        model = TestModel()
        x = torch.rand(2, 3, 5)
    
        with RewriterContext({}, onnx_custom_passes=_optimize_onnx):
            torch.onnx.export(
                model,
                x,
                onnx_file,
                input_names=['input'],
                output_names=['output1', 'output2'],
                dynamic_axes=dict(input=[0, 1, 2]))
    
        onnx_model = onnx.load(onnx_file)
        graph = onnx_model.graph
        nodes = graph.node
    
        node, idx = _find_next_node(0, nodes, 'Cast')
        assert node is not None
        assert node.attribute[0].i == 6
    
        node, idx = _find_next_node(idx + 1, nodes, 'Cast')
        assert node is not None
>       assert node.attribute[0].i == 1
E       assert 6 == 1
E        +  where 6 = name: "to"\ni: 6\ntype: INT\n.i

tests/test_apis/test_onnx_passes.py:124: AssertionError
____________________________________________ test_flatten_cls_head _____________________________________________

    def test_flatten_cls_head():
        pytest.importorskip('mmdeploy.backend.torchscript.ts_optimizer.onnx')
    
        try:
            from mmdeploy.backend.torchscript import ts_optimizer
            opt_pass = ts_optimizer.onnx._jit_pass_flatten_cls_head
        except ImportError:
            pytest.skip('pass not found.')
    
        def _optimize_onnx(graph, params_dict, torch_out):
            opt_pass(graph)
            return graph, params_dict, torch_out
    
        class TestModel(torch.nn.Module):
    
            def __init__(self) -> None:
                super().__init__()
    
            def forward(self, x):
                batch = x.size(0)
                gap = nn.functional.adaptive_avg_pool2d(x, (1, 1))
                gap = gap.reshape(batch, -1)
                return gap + 0  # gap should not be the output
    
        model = TestModel()
        x = torch.rand(1, 4, 8, 8)
    
        with RewriterContext({}, onnx_custom_passes=_optimize_onnx):
            torch.onnx.export(
                model,
                x,
                onnx_file,
                input_names=['input'],
                output_names=['output'],
                dynamic_axes=dict(input=[2, 3]))
    
        onnx_model = onnx.load(onnx_file)
        graph = onnx_model.graph
        nodes = graph.node
    
        node, idx = _find_next_node(0, nodes, 'GlobalAveragePool')
        assert node is not None
    
        node, idx = _find_next_node(idx + 1, nodes, 'Flatten')
>       assert node is not None
E       assert None is not None

tests/test_apis/test_onnx_passes.py:177: AssertionError
=============================================== warnings summary ===============================================
../../../anaconda3/envs/pt1.8/lib/python3.7/site-packages/torch/utils/cpp_extension.py:3
  /home/PJLAB/maningsheng/anaconda3/envs/pt1.8/lib/python3.7/site-packages/torch/utils/cpp_extension.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/test_apis/test_onnx_passes.py::test_merge_shape_concate
tests/test_apis/test_onnx_passes.py::test_peephole
tests/test_apis/test_onnx_passes.py::test_flatten_cls_head
  /home/PJLAB/maningsheng/anaconda3/envs/pt1.8/lib/python3.7/site-packages/torch/onnx/utils.py:1190: UserWarning: No names were found for specified dynamic axes of provided input.Automatically generated names will be applied to each dynamic axes of input input
    'Automatically generated names will be applied to each dynamic axes of input {}'.format(key))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================================== short test summary info ============================================
FAILED tests/test_apis/test_onnx_passes.py::test_merge_shape_concate - AssertionError: assert 'Constant' == '...
FAILED tests/test_apis/test_onnx_passes.py::test_peephole - assert 6 == 1
FAILED tests/test_apis/test_onnx_passes.py::test_flatten_cls_head - assert None is not None
=
```

File: tests/test_apis/test_onnx_passes.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            return x.new_zeros(x.shape[-2:])
+
+    model = TestModel()
+    x = torch.rand(1, 3, 4, 8)
+
+    with RewriterContext({}, onnx_custom_passes=_optimize_onnx):
+        torch.onnx.export(
+            model,
+            x,
+            onnx_file,
```
Comment (RunningLeon) [inline tests/test_apis/test_onnx_passes.py:52]: should we set `opset_version=11` here?

File: tests/test_apis/test_onnx_passes.py (hunk 2 above)
Comment (grimoire) [reply tests/test_apis/test_onnx_passes.py:3]: Which PyTorch version are you using?
Can you share the generated ONNX file here?

Comment (RunningLeon) [reply tests/test_apis/test_onnx_passes.py:3]: torch==1.8.0. Seems onnx passes are not executed.
[t.log](https://github.com/open-mmlab/mmdeploy/files/8795208/t.log)

Comment (grimoire) [reply tests/test_apis/test_onnx_passes.py:3]: Can not reproduce. @AllentDan would you mind take a look?

Comment (AllentDan) [reply tests/test_apis/test_onnx_passes.py:3]: In my testing yesterday, it failed too.

**Comment by RunningLeon** [comment]
may fix conflicts

File: tests/test_apis/test_onnx_passes.py (hunk 2 above)
Comment (grimoire) [reply tests/test_apis/test_onnx_passes.py:3]: > torch==1.8.0. Seems onnx passes are not executed. [t.log](https://github.com/open-mmlab/mmdeploy/files/8795208/t.log)

Yes, it's not. I guess the rewriter:
```python
from mmdeploy.apis.onnx.optimizer import \
    model_to_graph__custom_optimizer  # noqa
```
has not been trigered? Could you test if `model_to_graph__custom_optimizer` will replace `model_to_graph`?

Comment (RunningLeon) [reply tests/test_apis/test_onnx_passes.py:3]: Yes. It has been triggered. But it seems `opt_pass(graph)` does not change the `graph` at all.

**Comment by RunningLeon** [comment]
@grimoire Hi, does user has to install `libtorch` in order to benefit from this feature? If that's the case, we may have to add reminder in doc.

**Comment by RunningLeon** [comment]
@grimoire Hi, does user has to install `libtorch` in order to benefit from this feature? If that's the case, we may have to add reminder in doc.

**Comment by grimoire** [comment]
> @grimoire Hi, does user have to install `libtorch` in order to benefit from this feature? If that's the case, we may have to add reminder in doc.

They can also build with PyTorch just like mmcv. Which docs should I add?

**Comment by RunningLeon** [comment]
> > @grimoire Hi, does user have to install `libtorch` in order to benefit from this feature? If that's the case, we may have to add reminder in doc.
> 
> They can also build with PyTorch just like mmcv. Which docs should I add?

Maybe in a doc of  `onnx_optimizer` briefly introducing the experimental feature and the reminder?

**Review by RunningLeon — APPROVED** [review]
LGTM

**Review by AllentDan — APPROVED** [review]

**Comment by YOODS-Xu** [comment]
不好意思，请教一个非常初步的问题。
我在按照以下文档记载的办法安装时，报没有CMakeLists.txt的错误
https://github.com/open-mmlab/mmdeploy/blob/main/docs/en/experimental/onnx_optimizer.md

1.执行的命令
export Torch_DIR=$(python -c "import torch;print(torch.utils.cmake_prefix_path + '/Torch')")
cd $Torch_DIR
cmake \
    -DTorch_DIR=${Torch_DIR} \
    -DMMDEPLOY_TARGET_BACKENDS="onnxruntime;torchscript"

2.环境
Torch_DIR指向
miniconda3/envs/openmmlab/lib/python3.8/site-packages/torch/share/cmake/Torch

这个目录下只有
TorchConfig.cmake
TorchConfigVersion.cmake
这2个文件

3.问题
报没有CMakeLists.txt的错误

请问我怎么才能解决这个问题？
我的mmdeploy是用
pip install mmdeploy==1.3.1
安装的

**Pull request merged**